### DISCUSSION TITLE
Discord bot together with this

### DISCUSSION THREAD

**Discussion by everlyy** [ValvePython/steam#392] [Q&A] [answered]
I'm making a Discord bot that uses this library but I'm having some trouble doing the events.
I have a `friend_new` event set up, but it doesn't get called as soon as I get a friend request and I *think* it's because my code is waiting for Discord to process it's events, so the only time the `friend_new` event actually gets called is when I login to steam (either doing it in the code, or just restarting my program). 
I'd like to know how I can properly use a Discord bot together with this because I'm lost with the entire event thing.

Code looks like this:
```py
@friendslist.on("friend_invite")
def fl_friend_invite(user):
	print(f"New invite from {user.steam_id} ({user.name}), adding...")
	friendslist.add(user)

@friendslist.on("friend_new")
def fl_friend_new(user):
	print(f"New friend: {user.steam_id} ({user.name})!")

@client.event 
async def on_ready():
	print("[Discord] Client ready")
	login() # This logs in to steam

if __name__ == "__main__":
	client.run(discord_token, bot=True)
```

If you need more info to help me please ask because I'm completely lost.

**Comment by rossengeorgiev** [comment, marked as answer]
Yes, you need to give gevent loop a change to run. If you don't monekypatch, or call gevent cooperative function, the loop will never run. Mixing gevent and async and other, can be problematic. You should never try to monkey patch if you are mixing, because it will most likely break stuff. The only alternative is give a chance to each one to run. Since you are mainly running the discord client, setup a method that runs on a short interval and calls `client.idle()`. The call will block and run any outstanding gevent work and return. Avoid making any long blocking calls like computation or non-gevent cooperative io (e.g. unpatched `requests`).

**Reply by everlyy** [reply]
Yess, it works!! Thank you so much!!

For those wondering how I fixed it, I added a task to Discord's loop that runs `client.idle()` every 0.5 seconds.
```py
async def steam_idle():
	while True:
		sc.idle()
		await asyncio.sleep(.5)

@client.event 
async def on_ready():
	client.loop.create_task(steam_idle())
``` 

`client` is a discord client and `sc` is a steam client.

**Comment by truenicoco** [comment]
Hey, thanks for the tip. I did not understand how to use the gevent thing and found this topic.

I am getting a lot of noise like:

```

2022-08-12T19:34:05.100651713Z DEBUG:SteamClient:Emit event: <EMsg.Multi: 1>
2022-08-12T19:34:05.606735088Z DEBUG:SteamClient:Multi: Unpacking
2022-08-12T19:34:05.607222713Z DEBUG:SteamClient:Incoming: <MsgProto(<EMsg.ClientPersonaState: 766> | CMsgClientPersonaState)>
2022-08-12T19:34:05.607306047Z DEBUG:SteamClient:Emit event: <EMsg.ClientPersonaState: 766>
2022-08-12T19:34:05.607718005Z DEBUG:SteamClient:Incoming: <MsgProto(<EMsg.Multi: 1> | CMsgMulti)>
2022-08-12T19:34:05.608114213Z DEBUG:SteamClient:Emit event: <EMsg.Multi: 1>
2022-08-12T19:34:06.112408130Z DEBUG:SteamClient:Incoming: <MsgProto(<EMsg.Multi: 1> | CMsgMulti)>
2022-08-12T19:34:06.112547422Z DEBUG:SteamClient:Emit event: <EMsg.Multi: 1>
2022-08-12T19:34:06.615185339Z DEBUG:SteamClient:Multi: Unpacking
2022-08-12T19:34:06.615462172Z DEBUG:SteamClient:Incoming: <MsgProto(<EMsg.ClientPersonaState: 766> | CMsgClientPersonaState)>
2022-08-12T19:34:06.615570547Z DEBUG:SteamClient:Emit event: <EMsg.ClientPersonaState: 766>
2022-08-12T19:34:06.615652880Z DEBUG:SteamClient:Incoming: <MsgProto(<EMsg.Multi: 1> | CMsgMulti)>
2022-08-12T19:34:06.615685047Z DEBUG:SteamClient:Emit event: <EMsg.Multi: 1>
[...]
```

Is this expected?

**Reply by everlyy** [reply]
Seems like you have logging enabled in some way, it's all debug output

**Reply by truenicoco** [reply]
I have debug logging enabled indeed, but was surprised by the 'chatiness' of the library, even when idle. EMsg.Multi and EMsg.ClientPersonaState are used for heartbeats I guess? They seem unusually frequent to me, but maybe that's how the steam protocol works. I just want to avoid getting my account locked or whatever for making too many calls too often. We'll see :) Thanks for your reply.

**Reply by rossengeorgiev** [reply]
These are all `Incoming:` message from Steam.

`EMsg.ClientPersonaState` are updates when the status of one of your friends changes. If you have a lot of friends, and changing their status, especially game that provide rich presence (like what map, game mode, etc), then you get absolutely spammed with these message. It normal, and there's nothing you can do about it.

`EMsg.Multi` is message containing multiple messages as a optimization, instead of getting them one by one. Yes, Steam can send you a "multi" with just one message inside.
